components/places/android/src/main/java/org/mozilla/places/Bookmarks.kt

```diff
@@ -0,0 +1,407 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.places
+
+import java.lang.RuntimeException
+
+
+/**
+ * Enumeration of the ids of the roots of the bookmarks tree.
+ *
+ * There are 5 "roots" in the bookmark tree. The actual root
+ * (which have no parent), and it's 4 children (which have the
+ * actual root as their parent).
+ *
+ * You cannot delete or move any of these items.
+ */
+enum class BookmarkRoot(val id: String) {
+    Root("root________"),
+    Menu("menu________"),
+    Toolbar("toolbar_____"),
+    Unfiled("unfiled_____"),
+    Mobile("mobile______"),
+}
+
+/**
+ * Enumeration of the type of a bookmark item.
+ *
+ * Must match BookmarkType in the Rust code.
+ */
+enum class BookmarkType(val value: Int) {
+    Bookmark(1),
+    Folder(2),
+    Separator(3),
+}
+
+/**
+ * An interface defining the set of fields common to all nodes
+ * in the bookmark tree.
+ */
+interface BookmarkTreeNode {
+    /**
+     * The type of this bookmark.
+     */
+    val type: BookmarkType
+
+    /**
+     * The guid of this record. Bookmark guids are always 12 characters in the url-safe
+     * base64 character set.
+     */
+    val guid: String
+
+    /**
+     * Creation time, in milliseconds since the unix epoch.
+     *
+     * May not be a local timestamp.
+     */
+    val dateAdded: Long
+
+    /**
+     * Last modification time, in milliseconds since the unix epoch.
+     *
+     * May not be a local timestamp.
+     */
+    val lastModified: Long
+
+    /**
+     * The guid of this record's parent. It should only be null for
```

Comment:
nit: (unsure yet how this might fit w/ the rest of the api) You could enforce the "can't be null for non-root nodes" limitation at the interface level.

Reply:
I'd rather keep things simple than optimize for enforcing as much as possible, but if you have a concrete idea then LMK and I'll think about it.